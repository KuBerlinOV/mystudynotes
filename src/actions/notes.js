import { v4 as uuidv4 } from 'uuid';

//ADD_NOTE
export const addNote = ({ topic = '', // in this case you should imagine destructuring as you are passing some object with the same data like ' const movie = {title, year, id} '
    description = '',           //when we are passing it to the action function as an argument and destructuring it. 
    note = '',                 //when it would be like ' const {title, year, id} = movie ' 
    reference = '',            //Furthermore, we provide here the default values for the object keys so
    tag = '#',
    status = 'in progress',     //the user could leave these options empty if he wanted to.
    createdAt = 0
} = {}) => {
    return {
        type: 'ADD_NOTE',
        note: {
            id: uuidv4(),
            topic,
            description,
            note,
            reference,
            tag,
            status,
            createdAt
        }
    }
}

//EDIT_NOTE

export const updateNote = (id, updates) => ({
    type: 'UPDATE_NOTE',
    id,
    updates
})

export const updateStatus = (id) => ({
    type: 'UPDATE_STATUS',
    id
})

//REMOVE_NOTE

export const removeNote = (id) => ({
    type: 'REMOVE_NOTE',
    id
})