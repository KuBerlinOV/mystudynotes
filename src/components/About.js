import React from 'react';

const About = () => {
    return (
        <div>organizing your studies</div>
    )
}

export default About;