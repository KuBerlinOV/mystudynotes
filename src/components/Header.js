import React from 'react';
import { NavLink } from 'react-router-dom';


const Header = () => (
    <header>
        <NavLink activeClassName='is-active' className="nav-link" to="/" exact={true}>Home</NavLink>
        <NavLink activeClassName='is-active' className="nav-link" to="/notes">Notes</NavLink>
        <NavLink activeClassName='is-active' className="nav-link" to="/about">Organizing your studies</NavLink>
    </header>
)

export default Header;
