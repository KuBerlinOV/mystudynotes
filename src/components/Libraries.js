import React from 'react';

import Notes from './Notes'

const Libraries = () => {
    return (
        <div>
            <ul>
                <Notes />
            </ul>
        </div>
    )
};

export default Libraries