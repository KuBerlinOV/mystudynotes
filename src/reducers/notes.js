//reducers = notes
const notesReducerDefaultState = []

const notesReducer = (state = notesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_NOTE':
            return [...state, action.note];
        case 'UPDATE_NOTE':
            return state.map(note => {
                if (note.id === action.id) {
                    return {
                        ...note,
                        ...action.updates
                    }
                } else {
                    return note
                }
            });
        case 'REMOVE_NOTE':
            return state.filter(({ id }) => id !== action.id);
        case 'UPDATE_STATUS':
            return state.map(note => {
                if (note.id === action.id && note.status === 'in progress') {
                    return {
                        ...note,
                        status: 'mastered'
                    }
                } else if (note.id === action.id && note.status === 'mastered') {
                    return {
                        ...note,
                        status: 'in progress'
                    }
                } else {
                    return {
                        ...note
                    }
                }
            })
        default:
            return state;
    }

}

export default notesReducer;
