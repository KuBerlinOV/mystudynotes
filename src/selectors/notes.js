


const selectNotes = (notes, { text, sortBy, startDate, endDate }) => {
    return notes.filter(note => {
        const descriptionToMatch = note.description.toLowerCase()
        const topicToMatch = note.topic.toLowerCase();
        const textToMatch = text.toLowerCase();

        const textMatch = descriptionToMatch.includes(textToMatch) || topicToMatch.includes(textToMatch) ? true : false;
        const startDateMatch = typeof startDate !== 'number' || note.createdAt >= startDate;
        const endDateMatch = typeof endDate !== 'number' || note.createdAt <= endDate;

        return textMatch && startDateMatch && endDateMatch
    }).sort((a, b) => {
        if (sortBy === 'date') {
            return a.createdAt < b.createdAt ? 1 : -1
        } else if (sortBy === 'status') {
            return a.status === 'mastered' ? 1 : -1
        }
    })
}

export default selectNotes;