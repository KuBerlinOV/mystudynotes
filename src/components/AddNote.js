import React from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router';
import { addNote } from '../actions/notes';
import NoteForm from './NoteForm';

export const AddNote = (props) => {

    return (
        <NoteForm
            handleSubmit={(note) => {
                props.addNote(note);
                props.history.push('/notes');
            }}
        />
    )
}

const mapDispatchToProps = (dispatch) => ({
    addNote: (note) => dispatch(addNote(note))
})

export default connect(undefined, mapDispatchToProps)(AddNote);