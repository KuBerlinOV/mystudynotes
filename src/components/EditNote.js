import React from 'react';
import { connect } from 'react-redux';
import NoteForm from './NoteForm';
import { updateNote } from '../actions/notes';


export const EditNote = (props) => {
    return (
        <div>
            <NoteForm
                //note before update
                note={props.note}
                //this is handle submit props
                handleSubmit={(note) => {
                    //dispatch action to edit the expense
                    props.editNote(props.note.id, note);
                    //redirect to dashBoard
                    props.history.push('/notes')
                }}
            />
        </div>
    )
}

const mapDispatchToProps = (dispatch) => ({
    editNote: (id, note) => dispatch(updateNote(id, note))
})

const mapStateToProps = (state, props) => {
    return {
        note: state.notes.find((note) => note.id == props.match.params.id)
    }

}

export default connect(mapStateToProps, mapDispatchToProps)(EditNote);