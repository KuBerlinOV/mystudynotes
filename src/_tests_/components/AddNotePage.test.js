import React from 'react';
import { shallow } from 'enzyme';
import { AddNote } from '../../components/AddNote';
import notes from '../fixtures/notes';
import { render } from '@testing-library/react';

let addNoteSpy, historySpy, wrapper;

beforeEach(() => {
    addNoteSpy = jest.fn();
    historySpy = { push: jest.fn() };
    wrapper = shallow(<AddNote addNote={addNoteSpy} history={historySpy} />);
})


describe('AddNote', () => {
    it('should render AddNote correctly', () => {
        const { container } = render(<AddNote addNote={addNoteSpy} history={historySpy} />);
        expect(container).toMatchSnapshot();
    })
    //in this test enzyme is used in order to access the handleSubmit prop of the functional component
    //and test if the handleSubmit prop (function created in the NoteForm component) is passing down
    // the note object to dispatched addNote action
    it('should handle submit', () => {
        wrapper.find('NoteForm').prop('handleSubmit')(notes[1]) //<- I grabed the handleSubmit props and called it with the note object
        expect(historySpy.push).toHaveBeenLastCalledWith('/notes')
        expect(addNoteSpy).toHaveBeenLastCalledWith(notes[1])
    })
})