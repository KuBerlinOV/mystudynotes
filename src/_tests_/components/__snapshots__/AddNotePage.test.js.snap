// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddNote should render AddNote correctly 1`] = `
<div>
  <div>
    <form
      action=""
      data-testid="form"
    >
      <h3>
        Create Note
      </h3>
      <h3>
        Topic:
      </h3>
      <input
        name="topic"
        placeholder="topic"
        type="text"
        value=""
      />
      <h3>
        Description
      </h3>
      <input
        id=""
        name="description"
        placeholder="description"
        type="text"
        value=""
      />
      <h3>
        Note
      </h3>
      <textarea
        cols="30"
        id=""
        name="note"
        placeholder="write your note here"
        rows="10"
      />
      <h3>
        Reference
      </h3>
      <input
        id=""
        name="reference"
        placeholder="reference"
        type="text"
        value=""
      />
      <h3>
        Tag
      </h3>
      <input
        name="tag"
        placeholder="tag"
        type="text"
        value=""
      />
      <div>
        
        <button>
          Save
        </button>
      </div>
    </form>
  </div>
</div>
`;
